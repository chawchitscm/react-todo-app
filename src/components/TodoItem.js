import React, { useEffect, useState } from "react";

const TodoItem = props => {
  const [strike, setStrike] = useState("todo-text");

  useEffect(() => {
    if(props.clear) {
      setStrike("todo-text");
    }
    if (props.todo.checked) {
      setStrike("todo-text strikethrough");
    } else {
      setStrike("todo-text");
    }
  }, [props.clear, props.todo]);

  const handleCheck = () => {
    if (strike === "todo-text") {
      setStrike("todo-text strikethrough");
      props.handleCheck(props.id, true);
    } else {
      setStrike("todo-text");
      props.handleCheck(props.id, false);
    }
  }

  return (
    <div className="todo-item-wrapper">
      <input type="checkbox" onChange={handleCheck} checked={props.todo.checked } />
      <p className={strike}>{props.todo.text}</p>
      <button onClick={() => props.handleDelete(props.id)}>
        <svg xmlns="http://www.w3.org/2000/svg" width="18" height="18"><path fill="#494C6B" fillRule="evenodd" d="M16.97 0l.708.707L9.546 8.84l8.132 8.132-.707.707-8.132-8.132-8.132 8.132L0 16.97l8.132-8.132L0 .707.707 0 8.84 8.132 16.971 0z"/></svg>
      </button>
    </div>
  )
}

export default TodoItem;