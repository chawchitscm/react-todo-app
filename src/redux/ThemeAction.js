export const APPLY_THEME = "APPY_THEME";
export const applyTheme = (theme) => {
  const newTheme = theme === "dark" ? "light" : "dark";
  return {
    type: APPLY_THEME,
    payload: newTheme
  };
};