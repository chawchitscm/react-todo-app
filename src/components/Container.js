import React, { useEffect, useState } from "react";
import TodoItem from "./TodoItem";

const Container = () => {
  const [text, setText] = useState("");
  const [todos, setTodos] = useState([]);
  const [filteredTodos, setFilteredTodos] = useState([]);
  const [clear, setClear] = useState(false);
  const [type, setType] = useState("all");

  useEffect(() => {
    if (type === "all") {
      setFilteredTodos(todos);
    } else if (type === "active") {
      setFilteredTodos(todos.filter(todo => todo.checked === false));
    } else if (type === "completed") {
      setFilteredTodos(todos.filter(todo => todo.checked === true));
    }
  }, [type, todos]);

  const handleEnter = event => {
    if (event.keyCode === 13) {
      setTodos(todos => [...todos, {
        text: event.target.value,
        checked: false
      }]);
      setText("");
    }
  }

  const handleDelete = (id) => {
    var todosArray = [...todos];
    if (id !== -1) {
      todosArray.splice(id, 1);
      setTodos(todosArray);
    }
  }

  const handleCheck = (id, checkedParam) => {
    var todosArray = [...todos];
    var todo = {...todosArray[id]}
    if (checkedParam) {
      todo.checked = true;
      setClear(false);
    } else {
      todo.checked = false;
    }
    todosArray[id] = todo;
    setTodos(todosArray);
  }

  const handleClear = () => {
    setTodos(todos => todos.filter(todo => todo.checked === false));
    setClear(true);
  }

  const handleAll = () => {
    setType("all");
  }

  const handleActive = () => {
    setType("active");
  }

  const handleCompleted = () => {
    setType("completed");
  }

  return (
    <div className="container">
      <h1 className="todo-title">TODO</h1>
      <div className="input-wrapper">
        <input type="text" value={text} onChange={e => setText(e.target.value)} onKeyDown={handleEnter} />
      </div>
      <div className="todo-list">
      {
        filteredTodos.map((todo, index) => 
          <TodoItem 
            key={index}
            id={index}
            todo={todo}
            clear={clear}
            handleDelete={handleDelete}
            handleCheck={handleCheck} />
        )
      }
        <div className="todo-item-wrapper foot-nav">
          <p className="left-item">{todos.filter(todo => todo.checked === false).length} items left</p>
          <div className="action-btns">
            <button className={(type==="all" ? "active" : "")} onClick={handleAll}>All</button>
            <button className={(type==="active" ? "active" : "")} onClick={handleActive}>Active</button>
            <button className={(type==="completed" ? "active" : "")} onClick={handleCompleted}>Completed</button>
          </div>
          <button onClick={handleClear} className="clear-btn">Clear Completed</button>
        </div>
      </div>
    </div>
  )
}

export default Container;