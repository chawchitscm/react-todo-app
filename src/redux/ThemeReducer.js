import * as Actions from "./ThemeAction"
const initialState  = {
  theme: window.matchMedia("(prefers-color-scheme: light)").matches ? "light" : "dark"
}
export const ThemeReducer = (state=initialState, action) => {
  switch (action.type) {
    case Actions.APPLY_THEME:
      return Object.assign({}, {
        theme: action.payload
      });
    default:
      return state;
  }
};