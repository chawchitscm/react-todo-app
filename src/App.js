import "./App.css";
import Container from "./components/Container";
import useLocalStorage from "use-local-storage";

function App() {
  const defaultLight = window.matchMedia("(prefers-color-scheme: light)").matches;
  const [theme, setTheme] = useLocalStorage("theme", defaultLight ? "light" : "dark");

  const switchTheme = () => {
    const newTheme = theme === "dark" ? "light" : "dark";
    setTheme(newTheme);
  }
  return (
    <div className="App" data-theme={theme}>
      <Container 
        theme={theme}
        switchTheme={switchTheme} 
      />
    </div>
  );
}

export default App;
