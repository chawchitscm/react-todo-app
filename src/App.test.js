import "./matchMedia.mock";
import { render, screen, cleanup, fireEvent } from "@testing-library/react";
import renderer from "react-test-renderer";
import userEvent from '@testing-library/user-event';
import * as Reducer from "./redux/ThemeReducer";
import * as Actions from "./redux/ThemeAction";
import TodoItem from "./components/TodoItem";
import Container from "./components/Container";
import App from "./App";
import { shallow, configure, mount } from "enzyme";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";


afterEach(cleanup)

describe("Reducer Test", () => {
  test("should change to dark", () => {
    const initialState  = {
      theme: "light"
    }
    expect(Reducer.ThemeReducer(initialState, Actions.applyTheme(initialState.theme))).toEqual({
      theme: "dark"
    });
  });

  test("should change to light", () => {
    const initialState  = {
      theme: "dark"
    }
    expect(Reducer.ThemeReducer(initialState, Actions.applyTheme(initialState.theme))).toEqual({
      theme: "light"
    });
  });
});

describe("TodoItem Test", () => {
  test("show todo text from props", () => {
    // prepare
    const handleCheck = jest.fn();
    const handleDelete = jest.fn();
    const key_id = 1;
    // execute
    const todoItem = render(
      <TodoItem 
        key={key_id}
        id={key_id}
        todo={{
          text: "This is todo text.",
          checked: false
        }}
        clear={false}
        handleDelete={handleDelete}
        handleCheck={handleCheck} />
    );
    // assert
    expect(todoItem.getByText(/This is todo/i).textContent).toEqual("This is todo text.");
  });

  test("check action called on click", () => {
    // prepare
    const handleCheck = jest.fn();
    const handleDelete = jest.fn();
    const key_id = 1;
    // execute
    const { getAllByTestId } = render(
      <TodoItem 
        key={key_id}
        id={key_id}
        todo={{
          text: "This is todo text.",
          checked: false
        }}
        clear={false}
        handleDelete={handleDelete}
        handleCheck={handleCheck} />
    );
    const checkbox = getAllByTestId(key_id)[0];
    fireEvent.click(checkbox);
    // assert
    expect(checkbox.checked).toEqual(false);
    expect(handleCheck).toHaveBeenCalledTimes(1);
  });

  test("delete action called on click", () => {
    // prepare
    const handleCheck = jest.fn();
    const handleDelete = jest.fn();
    const key_id = 1;
    // execute
    const { getAllByTestId } = render(
      <TodoItem 
        key={key_id}
        id={key_id}
        todo={{
          text: "This is todo text.",
          checked: false
        }}
        clear={false}
        handleDelete={handleDelete}
        handleCheck={handleCheck} />
    );
    const deleteBtn = getAllByTestId(key_id)[1];
    fireEvent.click(deleteBtn);
    // assert
    expect(handleDelete).toHaveBeenCalledTimes(1);
  });
});

describe("Container Test", () => {
  test("switch theme on click", () => {
    // prepare
    const switchTheme = jest.fn();
    const theme = "light"
    // execute
    const {getByTestId} = render(
      <Container 
        theme={theme}
        switchTheme={switchTheme} />
    );
    const themeSwitchBtn = getByTestId("switch");
    fireEvent.click(themeSwitchBtn);
    // assert
    expect(switchTheme).toHaveBeenCalledTimes(1);
  });

  test("add new todo on enter", () => {
    // prepare
    const container = render(<Container />);
    const todoInput = container.getByTestId("todo-input");
    // execute
    userEvent.type(todoInput, "New Todo");
    // assert
    expect(screen.getByTestId("todo-input")).toHaveValue("New Todo");
    // execute
    fireEvent.keyDown(todoInput, {
      keyCode: 13
    });
    // assert
    expect(container.getByText(/New/i).textContent).toEqual("New Todo");
  });

  test("remove the component on clicking delete", () => {
    // prepare
    const container = render(<Container />);
    const todoInput = container.getByTestId("todo-input");
    // execute
    userEvent.type(todoInput, "New Todo");
    // assert
    expect(screen.getByTestId("todo-input")).toHaveValue("New Todo");
    // execute
    fireEvent.keyDown(todoInput, {
      keyCode: 13
    });
    // assert
    expect(container.getByText(/New/i).textContent).toEqual("New Todo");
    const deleteBtn = container.getAllByTestId(0)[1];
    // execute
    fireEvent.click(deleteBtn);
    // assert
    expect(deleteBtn).not.toBeInTheDocument();
  });

  test("checkbox checked on click", () => {
    // prepare
    const key_id = 0;
    const container = render(<Container />);
    const todoInput = container.getByTestId("todo-input");
    // execute
    userEvent.type(todoInput, "New Todo");
    // assert
    expect(screen.getByTestId("todo-input")).toHaveValue("New Todo");
    // execute
    fireEvent.keyDown(todoInput, {
      keyCode: 13
    });
    const checkbox = container.getAllByTestId(key_id)[0];
    // assert
    expect(checkbox.checked).toEqual(false);
    // execute
    fireEvent.click(checkbox);
    // assert
    expect(checkbox.checked).toEqual(true);
  });

  test("remove completed components on clicking clear", () => {
    // prepare
    const key_id = 0;
    const container = render(<Container />);
    const todoInput = container.getByTestId("todo-input");
    // execute
    userEvent.type(todoInput, "New Todo");
    // assert
    expect(screen.getByTestId("todo-input")).toHaveValue("New Todo");
    // execute
    fireEvent.keyDown(todoInput, {
      keyCode: 13
    });
    const checkbox = container.getAllByTestId(key_id)[0];
    const deleteBtn = container.getAllByTestId(key_id)[1];
    // assert
    expect(checkbox.checked).toEqual(false);
    // execute
    fireEvent.click(checkbox);
    // assert
    expect(checkbox.checked).toEqual(true);
    const clearBtn = container.getByTestId("clear");
    // execute
    fireEvent.click(clearBtn);
    // assert
    expect(deleteBtn).not.toBeInTheDocument();
  });

  test("matches snapchost", () => {
    // create virtual dom
    const tree = renderer.create(<Container/>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("show all todos", () => {
    // prepare
    const container = render(<Container />);
    const todoInput = container.getByTestId("todo-input");
    userEvent.type(todoInput, "Checked Todo");
    fireEvent.keyDown(todoInput, {
      keyCode: 13
    });
    const checkbox = container.getAllByTestId(0)[0];
    fireEvent.click(checkbox);
    userEvent.type(todoInput, "Active Todo");
    fireEvent.keyDown(todoInput, {
      keyCode: 13
    });
    const allBtn = container.getByTestId("all");
    // execute
    fireEvent.click(allBtn);
    // assert
    expect(container.getByText(/Checked Todo/i).textContent).toEqual("Checked Todo");
    expect(container.getByText(/Active Todo/i).textContent).toEqual("Active Todo");
  });

  test("show active todos", () => {
    // prepare
    const container = render(<Container />);
    const todoInput = container.getByTestId("todo-input");
    userEvent.type(todoInput, "Todo Checked");
    fireEvent.keyDown(todoInput, {
      keyCode: 13
    });
    const checkbox = container.getAllByTestId(0)[0];
    fireEvent.click(checkbox);
    userEvent.type(todoInput, "Todo Active");
    fireEvent.keyDown(todoInput, {
      keyCode: 13
    });
    const activeBtn = container.getByTestId("active");
    // execute
    fireEvent.click(activeBtn);
    const checkedTodo = screen.queryAllByText("Todo Checked")
    // assert
    expect(container.getByText(/Todo Active/i).textContent).toEqual("Todo Active");
    expect(checkedTodo).toHaveLength(0);
  });

  test("show completed todos", () => {
    // prepare
    const container = render(<Container />);
    const todoInput = container.getByTestId("todo-input");
    userEvent.type(todoInput, "Todo Checked");
    fireEvent.keyDown(todoInput, {
      keyCode: 13
    });
    const checkbox = container.getAllByTestId(0)[0];
    fireEvent.click(checkbox);
    userEvent.type(todoInput, "Todo Active");
    fireEvent.keyDown(todoInput, {
      keyCode: 13
    });
    const activeBtn = container.getByTestId("completed");
    // execute
    fireEvent.click(activeBtn);
    const checkedTodo = screen.queryAllByText("Todo Active")
    // assert
    expect(container.getByText(/Todo Checked/i).textContent).toEqual("Todo Checked");
    expect(checkedTodo).toHaveLength(0);
  });

  test("double check undo the check state", () => {
    // prepare
    const container = render(<Container />);
    const todoInput = container.getByTestId("todo-input");
    userEvent.type(todoInput, "Todo Checked");
    fireEvent.keyDown(todoInput, {
      keyCode: 13
    });
    const checkbox = container.getAllByTestId(0)[0];
    fireEvent.click(checkbox);
    expect(checkbox.checked).toEqual(true);
    fireEvent.click(checkbox);
    expect(checkbox.checked).toEqual(false);
  })
});

configure({ adapter: new Adapter() });
describe("App", () => {
  const initialState  = {
    theme: "light"
  }
  const mockStore = configureStore();
  let store;
  it("renders correctly", () => {
    store = mockStore(initialState);
    shallow(<Provider store={store}><App /></Provider>);
  });

  it("include one h1", () => {
    store = mockStore(initialState);
    const wrapper = mount(<Provider store={store}><App /></Provider>);
    expect(wrapper.find("h1").length).toEqual(1)
  });

  it("called switch theme", () => {
    // const switchTheme = jest.fn();
    store = mockStore(initialState);
    const wrapper = mount(<Provider store={store}><App /></Provider>);
    // const instance = wrapper.instance();
    // const spy = jest.spyOn(instance, "switchTheme");
    // instance.forceUpdate();
    // expect(spy).toHaveBeenCalled();
    wrapper.find("#switch-btn").simulate("click");
    // expect(spy).toHaveBeenCalled();
  });
});