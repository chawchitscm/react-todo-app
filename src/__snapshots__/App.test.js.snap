// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container Test matches snapchost 1`] = `
<div
  className="container"
>
  <div
    className="header"
  >
    <h1
      className="todo-title"
    >
      TODO
    </h1>
    <button
      data-testid="switch"
      id="switch-btn"
    >
      <svg
        height="26"
        width="26"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M13 21a1 1 0 011 1v3a1 1 0 11-2 0v-3a1 1 0 011-1zm-5.657-2.343a1 1 0 010 1.414l-2.121 2.121a1 1 0 01-1.414-1.414l2.12-2.121a1 1 0 011.415 0zm12.728 0l2.121 2.121a1 1 0 01-1.414 1.414l-2.121-2.12a1 1 0 011.414-1.415zM13 8a5 5 0 110 10 5 5 0 010-10zm12 4a1 1 0 110 2h-3a1 1 0 110-2h3zM4 12a1 1 0 110 2H1a1 1 0 110-2h3zm18.192-8.192a1 1 0 010 1.414l-2.12 2.121a1 1 0 01-1.415-1.414l2.121-2.121a1 1 0 011.414 0zm-16.97 0l2.121 2.12A1 1 0 015.93 7.344L3.808 5.222a1 1 0 011.414-1.414zM13 0a1 1 0 011 1v3a1 1 0 11-2 0V1a1 1 0 011-1z"
          fill="#FFF"
          fillRule="evenodd"
        />
      </svg>
    </button>
  </div>
  <div
    className="input-wrapper"
  >
    <input
      data-testid="todo-input"
      onChange={[Function]}
      onKeyDown={[Function]}
      type="text"
      value=""
    />
  </div>
  <div
    className="todo-list"
  >
    <div
      className="todo-item-wrapper foot-nav"
    >
      <p
        className="left-item"
      >
        0
         items left
      </p>
      <div
        className="action-btns"
      >
        <button
          className="active"
          data-testid="all"
          onClick={[Function]}
        >
          All
        </button>
        <button
          className=""
          data-testid="active"
          onClick={[Function]}
        >
          Active
        </button>
        <button
          className=""
          data-testid="completed"
          onClick={[Function]}
        >
          Completed
        </button>
      </div>
      <button
        className="clear-btn"
        data-testid="clear"
        onClick={[Function]}
      >
        Clear Completed
      </button>
    </div>
  </div>
</div>
`;
